lib/inventory/metadata.go

```diff
@@ -0,0 +1,233 @@
+// standard library. Having these two methods configurable allows us to mock
+// them in tests.
+func (c *fetchConfig) setDefaults() {
+	if c.readFile == nil {
+		c.readFile = os.ReadFile
+	}
+	if c.execCommand == nil {
+		c.execCommand = func(name string, args ...string) ([]byte, error) {
+			return exec.Command(name, args...).Output()
+		}
+	}
+	if c.kubeClient == nil {
+		c.kubeClient = getKubeClient()
+	}
+}
+
+// getKubeClient returns a kubernetes client in case the agent is running on
+// kubernetes. It returns nil otherwise.
+func getKubeClient() kubernetes.Interface {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		log.Debugf("Failed to get kubernetes cluster config: %s", err)
+		return nil
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Debugf("Failed to create kubernetes client: %s", err)
+		return nil
+	}
+	return client
+}
+
+// fetchAgentMetadata fetches and calculates all agent metadata we are interested
+// in tracking.
+func fetchAgentMetadata(c *fetchConfig) proto.UpstreamInventoryAgentMetadata {
+	c.setDefaults()
+	return proto.UpstreamInventoryAgentMetadata{
+		Version:               c.fetchVersion(),
+		HostID:                c.fetchHostID(),
+		Services:              c.fetchServices(),
+		OS:                    c.fetchOS(),
+		OSVersion:             c.fetchOSVersion(),
+		HostArchitecture:      c.fetchHostArchitecture(),
+		GLibCVersion:          c.fetchGlibcVersion(),
+		InstallMethods:        c.fetchInstallMethods(),
+		ContainerRuntime:      c.fetchContainerRuntime(),
+		ContainerOrchestrator: c.fetchContainerOrchestrator(),
+		CloudEnvironment:      c.fetchCloudEnvironment(),
+	}
+}
+
+// fetchVersion returns the Teleport version present in the hello message.
+func (c *fetchConfig) fetchVersion() string {
+	return c.hello.Version
+}
+
+// fetchHostID returns the agent ID present in the hello message.
+func (c *fetchConfig) fetchHostID() string {
+	return c.hello.ServerID
+}
+
+// fetchOS returns the value of GOOS.
+func (c *fetchConfig) fetchOS() string {
+	return runtime.GOOS
+}
+
+// fetchServices computes the list of access protocols enabled at the agent from
+// the list of system roles present in the hello message.
+func (c *fetchConfig) fetchServices() []string {
+	var services []string
+	for _, svc := range c.hello.Services {
+		switch svc {
+		case types.RoleNode:
+			services = append(services, nodeService)
+		case types.RoleKube:
+			services = append(services, kubeService)
+		case types.RoleApp:
+			services = append(services, appService)
+		case types.RoleDatabase:
+			services = append(services, dbService)
+		case types.RoleWindowsDesktop:
+			services = append(services, windowsDesktopService)
+		}
+	}
+	return services
+}
+
+// fetchHostArchitecture computes the host architecture using the arch
+// command-line utility.
+func (c *fetchConfig) fetchHostArchitecture() string {
+	command := "arch"
+	out, err := c.exec(command)
```

Comment:
Can `arch` ever return a value that's different than what we can infer from `runtime.GOARCH`? On macOS, at least, it seems to output `i386` when called from within a Rosetta-enabled binary.

Reply:
Yes. This was the motivation behind using `arch` instead of `runtime.GOARCH`.